chart/templates/configmap-env.yaml

```diff
@@ -0,0 +1,66 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "mastodon.fullname" . }}-env
+  labels:
+    {{- include "mastodon.labels" . | nindent 4 }}
+data:
+  DB_HOST: {{ template "mastodon.postgresql.fullname" . }}
+  DB_NAME: {{ .Values.postgresql.postgresqlDatabase }}
+  # https://github.com/tootsuite/mastodon/blob/master/dist/mastodon-sidekiq.service
+  DB_POOL: "25"
+  DB_PORT: "5432"
+  DB_USER: {{ .Values.postgresql.postgresqlUsername }}
+  DEFAULT_LOCALE: {{ .Values.locale }}
+  {{- if .Values.elasticsearch.enabled }}
+  ES_ENABLED: "true"
+  ES_HOST: {{ template "mastodon.elasticsearch.fullname" . }}-master
+  ES_PORT: "9200"
+  {{- end }}
+  LOCAL_DOMAIN: {{ .Values.ingress.hostname }}
+  # https://devcenter.heroku.com/articles/tuning-glibc-memory-behavior
+  MALLOC_ARENA_MAX: "2"
+  NODE_ENV: "production"
+  RAILS_ENV: "production"
+  REDIS_HOST: {{ template "mastodon.redis.fullname" . }}-master
+  REDIS_PORT: "6379"
+  {{- if .Values.smtp.auth_method }}
+  SMTP_AUTH_METHOD: {{ .Values.smtp.auth_method }}
+  {{- end }}
+  {{- if .Values.smtp.ca_file }}
+  SMTP_CA_FILE: {{ .Values.smtp.ca_file }}
+  {{- end }}
+  {{- if .Values.smtp.delivery_method }}
+  SMTP_DELIVERY_METHOD: {{ .Values.smtp.delivery_method }}
+  {{- end }}
+  {{- if .Values.smtp.domain }}
+  SMTP_DOMAIN: {{ .Values.smtp.domain }}
+  {{- end }}
+  {{- if .Values.smtp.enable_starttls_auto }}
+  SMTP_ENABLE_STARTTLS_AUTO: {{ .Values.smtp.enable_starttls_auto | quote }}
+  {{- end }}
+  {{- if .Values.smtp.from_address }}
+  SMTP_FROM_ADDRESS: {{ .Values.smtp.from_address }}
+  {{- end }}
+  {{- if .Values.smtp.login }}
+  SMTP_LOGIN: {{ .Values.smtp.login }}
+  {{- end }}
+  {{- if .Values.smtp.openssl_verify_mode }}
+  SMTP_OPENSSL_VERIFY_MODE: {{ .Values.smtp.openssl_verify_mode }}
+  {{- end }}
+  {{- if .Values.smtp.password }}
+  SMTP_PASSWORD: {{ .Values.smtp.password }}
+  {{- end }}
+  {{- if .Values.smtp.port }}
+  SMTP_PORT: {{ .Values.smtp.port | quote }}
+  {{- end }}
+  {{- if .Values.smtp.reply_to }}
+  SMTP_REPLY_TO: {{ .Values.smtp.reply_to }}
+  {{- end }}
+  {{- if .Values.smtp.server }}
+  SMTP_SERVER: {{ .Values.smtp.server }}
+  {{- end }}
+  {{- if .Values.smtp.tls }}
+  SMTP_TLS: {{ .Values.smtp.tls | quote }}
+  {{- end }}
+  STREAMING_CLUSTER_NUM: {{ .Values.streaming.workers | quote }}
```

Comment:
Instead of hand mapping (and sometime hard coding) these values, I would recommend that any that are not derived by mapped directly (so STREAMING_CLUSTER_NUM here is set by a STREAMING_CLUSTER_NUM over in values. This really simplicities both the conceptual model and the upkeep as new variables are added or needed. 

I have template code for this I can drop in Monday if that sounds interesting.

Reply:
I'd like to see your template for that, thanks!

I initially did allow adding arbitrary env values, and injected them all into the configmap with something like:
```
  {{- range .Values.env_vars }}
  {{ .name }}: {{ .value | quote }}
  {{- end }}
```
But I changed to hand mapping because I was afraid people would add all the values they thought they needed, without checking if they were already defined in the configmap.

Reply:
I'm suggesting removing everywhere that takes a structure in values.template and simply remaps in in to ENV vars for the containers and move it in to a more 1-1 structure with code like this:

```
{{- range $key, $value := .Values.env.configmap }}
  {{ $key }}: "{{ $value }}"
{{- end }}
```

In the values file you'd have things like 

```
env:
  configmap: 
    SMTP_AUTH_METHOD: 
    SMTP_CA_FILE:
    STREAMING_CLUSTER_NUM:
 ```
etc.  So you are still showing all the important ones in the template, but if you miss one they don't need a new chart version to add it. Also adding one to the chart example is as simple as updating the values template in one spot. 

Note - all the derived values I would keep as is.

Reply:
By derived values I assume you mean ones like
```
  DB_HOST: {{ template "mastodon.postgresql.fullname" . }}
  DB_NAME: {{ .Values.postgresql.postgresqlDatabase }}
```
since golang templating doesn't work in `values.yaml` and throws errors like
```
Error: cannot load values.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".Values.postgresql.postgresqlDatabase":interface {}(nil)}
```
But if we keep all those values where they are in the configmap and move the rest to `values.yaml` as their ENV names, then we get this: https://github.com/dunn/mastodon/commit/33e2c10a8f632a434b2cf4d1c4d7d10a147d5496

Which IMO isn't great, since anyone deploying the chart has to know to look in the configmap to avoid clobbering already-defined variables.  That's what I was trying to avoid with the hand-mapping approach.

Reply:
A minor thing, but we also lose the ability to flag required values in `secrets.yaml`, since we can't interpolate values into the error message; this
```
  {{- if (empty $value) }}
  {{ $key }}: {{ required "$key is required" $value }}
  {{- else }}
```
doesn't print a useful warning when the user fails to set a value for a given variable:
```
Error: UPGRADE FAILED: execution error at (mastodon/templates/secrets.yaml:11:17): $key is required
```